### PR TITLE
optimize hash cache data file writes

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

mmap the `whole` hash cache data file into memory to write, then `unmmap` the data file is not efficient. It increases the TLB pressure (more page table entries) and leads to higher dirty page ratio, which puts stress on the memory paging table.

Since writing out the hash cache data is sequential, a buffered file writer will put less stress on memory paging table resources and lead to better performance. 

- master (mmap, element by element writes)

```
sol@dev-equinix-washington-23:~$ grep cache_hash_data_stats  d1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "write_to_mmap_us"
write_to_mmap_us           28268583i
write_to_mmap_us           288182i
```

- mmap, slice writes
```
sol@dev-equinix-washington-23:~$ grep cache_hash_data_stats  d1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "write_to_mmap_us"
write_to_mmap_us           31239824i
write_to_mmap_us           305127i
```

It is slower to copy the slice (probably due to the fact that faster write lead to higher dirty page ratio and puts more pressure on memory paging).

- FileBufWriter

```
sol@dev-equinix-washington-23:~$ grep cache_hash_data_stats  d1 | awk '{for(i=1; i<=NF; i++) if (0 != index($i, "=")) {print($i)}}' | column -ts "=" | grep -E "write_to_mmap_us"                                                                            
write_to_mmap_us           22873415i
write_to_mmap_us           262826i
```

FileBufWriter saves 6s for writing out cached hash data into file. 

#### Summary of Changes

Use BufWriter to write out cached hash data file. 

Fixes #

### CODE REVIEW

File: accounts-db/src/cache_hash_data.rs [hunk 1]
```diff
@@ -363,13 +329,12 @@ impl CacheHashData {
         let mut m2 = Measure::start("write_to_mmap");
         let mut i = 0;
         data.iter().for_each(|x| {
-            x.iter().for_each(|item| {
-                let d = cache_file.get_mut(i as u64);
-                i += 1;
-                *d = item.clone();
-            })
+            let size = x.len() * std::mem::size_of::<EntryType>();
+            let slice = unsafe { std::slice::from_raw_parts(x.as_ptr() as *const u8, size) };
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs:333]: It seesm that to use bytemuck. We will need to add Pod traits for `CalculateHashIntermediate` and `sdk/program/Hash` - which may break the ABI?

File: accounts-db/src/cache_hash_data.rs [hunk 2]
```diff
@@ -339,19 +298,26 @@ impl CacheHashData {
         let entries = entries.iter().sum::<usize>();
         let capacity = cell_size * (entries as u64) + std::mem::size_of::<Header>() as u64;
 
-        let mmap = CacheHashDataFile::new_map(&cache_path, capacity)?;
+        let file = OpenOptions::new()
+            .create_new(true)
+            .write(true)
+            .append(true)
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: fixed.

File: accounts-db/src/cache_hash_data.rs [hunk 3]
```diff
@@ -339,19 +298,26 @@ impl CacheHashData {
...
+            .open(cache_path)
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: fixed.

File: accounts-db/src/cache_hash_data.rs [hunk 4]
```diff
@@ -339,19 +298,26 @@ impl CacheHashData {
...
+        let mut fw = std::io::BufWriter::new(file);
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: fixed.

File: accounts-db/src/cache_hash_data.rs [hunk 5]
```diff
@@ -339,19 +298,26 @@ impl CacheHashData {
...
+
+        let header = Header { count: entries };
+        let header_slice = unsafe {
+            std::slice::from_raw_parts(
+                (&header as *const Header) as *const u8,
+                std::mem::size_of::<Header>(),
+            )
+        };
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: fixed.

File: accounts-db/src/cache_hash_data.rs [hunk 6]
```diff
@@ -363,13 +329,12 @@ impl CacheHashData {
...
         let mut i = 0;
         data.iter().for_each(|x| {
-            x.iter().for_each(|item| {
-                let d = cache_file.get_mut(i as u64);
-                i += 1;
-                *d = item.clone();
-            })
+            let size = x.len() * std::mem::size_of::<EntryType>();
+            let slice = unsafe { std::slice::from_raw_parts(x.as_ptr() as *const u8, size) };
+            fw.write_all(slice).unwrap();
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: fixed.

File: accounts-db/src/cache_hash_data.rs [hunk 7]
```diff
@@ -363,13 +329,12 @@ impl CacheHashData {
...
+            i += x.len() as u64;
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: fixed.

File: accounts-db/src/cache_hash_data.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs]: updated.

File: sdk/program/src/hash.rs [hunk 8]
```diff
@@ -41,6 +42,8 @@ const MAX_BASE58_LEN: usize = 44;
     Ord,
     PartialOrd,
     Hash,
+    Pod,
+    Zeroable,
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply sdk/program/src/hash.rs:46]: Great. Thanks for confirming.

File: accounts-db/src/accounts_db.rs [hunk 9]
```diff
@@ -10661,10 +10661,10 @@ pub mod tests {
...
             .unwrap();
         let mut expected = vec![Vec::new(); bins];
-        expected[0].push(raw_expected[0].clone());
-        expected[0].push(raw_expected[1].clone());
-        expected[bins - 1].push(raw_expected[2].clone());
-        expected[bins - 1].push(raw_expected[3].clone());
+        expected[0].push(raw_expected[0]);
+        expected[0].push(raw_expected[1]);
+        expected[bins - 1].push(raw_expected[2]);
+        expected[bins - 1].push(raw_expected[3]);
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:10667]: Yup. Pod require Copy.

File: accounts-db/src/cache_hash_data.rs [hunk 10]
```diff
@@ -339,19 +300,20 @@ impl CacheHashData {
         let entries = entries.iter().sum::<usize>();
         let capacity = cell_size * (entries as u64) + std::mem::size_of::<Header>() as u64;
 
-        let mmap = CacheHashDataFile::new_map(&cache_path, capacity)?;
+        let file = OpenOptions::new()
+            .create_new(true)
+            .write(true)
+            .open(cache_path)?;
+        let mut fw = BufWriter::new(file);
```
Comment (jeffwashington) [inline accounts-db/src/cache_hash_data.rs:307]: I guess we should probably not go this route now?

Comment (HaoranYi) [reply accounts-db/src/cache_hash_data.rs:307]: yeah. closed.

**Pull request closed without merging**